crates/bitwarden/src/auth/password/validate.rs

```diff
@@ -34,8 +34,49 @@ pub(crate) async fn validate_password(
     }
 }
 
+#[cfg(feature = "internal")]
+pub(crate) fn validate_password_user_key(
+    client: &Client,
+    password: String,
+    encrypted_user_key: String,
+) -> Result<String> {
+    let login_method = client
+        .login_method
+        .as_ref()
+        .ok_or(Error::NotAuthenticated)?;
+
+    if let LoginMethod::User(login_method) = login_method {
+        match login_method {
+            UserLoginMethod::Username { email, kdf, .. }
+            | UserLoginMethod::ApiKey { email, kdf, .. } => {
+                let master_key = MasterKey::derive(password.as_bytes(), email.as_bytes(), kdf)?;
+                let user_key = master_key
+                    .decrypt_user_key(encrypted_user_key.parse()?)
+                    .map_err(|_| "wrong password")?;
+
+                let enc = client
+                    .get_encryption_settings()
+                    .map_err(|_| Error::VaultLocked)?;
+
+                let existing_key = enc.get_key(&None).ok_or(Error::VaultLocked)?;
+
+                if user_key.to_vec() != existing_key.to_vec() {
+                    return Err("wrong password".into());
```

Comment:
I think this error would be better as "wrong user key" or something like that, if the password was wrong we'd fail when decrypting already.